contracts/upgradability/diamond/diamond.rs

```diff
@@ -0,0 +1,221 @@
+        self._emit_diamond_cut_event(&diamond_cut, &init);
+
+        if init.is_some() {
+            self.flush();
+            self._init_call(init.unwrap());
+        }
+
+        Ok(())
+    }
+
+    default fn _fallback(&self) -> ! {
+        let selector = ink_env::decode_input::<Selector>().unwrap_or_else(|_| panic!("Calldata error"));
+
+        let delegate_code = self.get().selector_to_hash.get(selector);
+
+        if delegate_code.is_none() {
+            panic!("Function is not registered");
+        }
+
+        ink_env::call::build_call::<ink_env::DefaultEnvironment>()
+            .call_type(DelegateCall::new().code_hash(delegate_code.unwrap()))
+            .call_flags(
+                ink_env::CallFlags::default()
+                // We don't plan to use the input data after the delegated call, so the 
+                // input data can be forwarded to delegated contract to reduce the gas usage.
+                .set_forward_input(true)
+                // We don't plan to return back to that contract after execution, so we 
+                // marked delegated call as "tail", to end the execution of the contract.
+                .set_tail_call(true),
+            )
+            .fire()
+            .unwrap_or_else(|err| panic!("delegate call to {:?} failed due to {:?}", delegate_code, err));
+        unreachable!("the _fallback call will never return since `tail_call` was set");
+    }
+
+    default fn _init_call(&self, call: InitCall) -> ! {
+        ink_env::call::build_call::<ink_env::DefaultEnvironment>()
+            .call_type(DelegateCall::new().code_hash(call.hash))
+            .exec_input(ExecutionInput::new(InkSelector::new(call.selector)).push_arg(call.input))
+            .call_flags(ink_env::CallFlags::default()
+            // We don't plan to return back to that contract after execution, so we
+            // marked delegated call as "tail", to end the execution of the contract.
+            .set_tail_call(true))
+            .returns::<()>()
+            .fire()
+            .unwrap_or_else(|err| panic!("init call failed due to {:?}", err));
+        unreachable!("the _init_call call will never return since `tail_call` was set");
+    }
+
+    default fn _handle_replace_immutable(&mut self, hash: Hash) -> Result<(), DiamondError> {
+        return if hash == self.get().self_hash {
+            Err(DiamondError::ImmutableFunction)
+        } else {
+            Ok(())
+        }
+    }
+
+    default fn _remove_facet(&mut self, code_hash: Hash) {
+        let vec = self.get().hash_to_selectors.get(&code_hash).unwrap();
+        vec.iter().for_each(|old_selector| {
+            self.get_mut().selector_to_hash.remove(&old_selector);
+        });
+        self.get_mut().hash_to_selectors.remove(&code_hash);
+        self._on_remove_facet(code_hash);
+    }
+
+    default fn _remove_selectors(&mut self, facet_cut: &FacetCut) {
+        let selectors = self.get().hash_to_selectors.get(&facet_cut.hash).unwrap();
```

Comment:
You still need to update `hash_to_selectors` after removing of unused functions=)